File: libs/encryption/hmac/src/main/java/com/getcode/crypt/Hmac.java
package com.getcode.crypt;


import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class Hmac {
    private void generateHashWithHmac256(String message, String key) {
        try {
            final String hashingAlgorithm = "HmacSHA256";

            byte[] bytes = hmac(hashingAlgorithm, key.getBytes(), message.getBytes());

            final String messageDigest = bytesToHex(bytes);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static byte[] hmac(String algorithm, byte[] key, byte[] message) throws NoSuchAlgorithmException, InvalidKeyException {
        Mac mac = Mac.getInstance(algorithm);
        mac.init(new SecretKeySpec(key, algorithm));
        return mac.doFinal(message);
    }

    public static String bytesToHex(byte[] bytes) {
        final char[] hexArray = "0123456789abcdef".toCharArray();
        char[] hexChars = new char[bytes.length * 2];
        for (int j = 0, v; j < bytes.length; j++) {
            v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }
        return new String(hexChars);
    }
}

File: libs/encryption/hmac/src/commonMain/kotlin/com/getcode/crypt/Hmac.kt
package com.getcode.crypt

import org.kotlincrypto.macs.hmac.sha2.HmacSHA256
import org.kotlincrypto.macs.hmac.sha2.HmacSHA512

/** Multiplatform HMAC utilities backed by kotlincrypto. */
object Hmac {

    /**
     * Computes an HMAC over [message] using [key] with the given [algorithm].
     *
     * Supported algorithm strings: `"HmacSHA512"`, `"HmacSHA256"`.
     */
    fun hmac(algorithm: String, key: ByteArray, message: ByteArray): ByteArray {
        return when (algorithm) {
            "HmacSHA512" -> {
                val mac = HmacSHA512(key)
                mac.update(message)
                mac.doFinal()
            }
            "HmacSHA256" -> {
                val mac = HmacSHA256(key)
                mac.update(message)
                mac.doFinal()
            }
            else -> throw IllegalArgumentException("Unsupported HMAC algorithm: $algorithm")
        }
    }

    /** Encodes [bytes] as a lowercase hex string. */
    fun bytesToHex(bytes: ByteArray): String {
        val hexChars = CharArray(bytes.size * 2)
        for (j in bytes.indices) {
            val v = bytes[j].toInt() and 0xFF
            hexChars[j * 2] = "0123456789abcdef"[v ushr 4]
            hexChars[j * 2 + 1] = "0123456789abcdef"[v and 0x0F]
        }
        return hexChars.concatToString()
    }
}
